Auxiliary/DirectXTexJPEG.cpp

```diff
@@ -0,0 +1,374 @@
+            ShiftPixels(image);
+        #endif
+            return S_OK;
+        }
+
+        HRESULT GetImage(ScratchImage& image) noexcept(false)
+        {
+            TexMetadata metadata{};
+            return GetImage(metadata, image);
+        }
+    };
+
+    class JPEGCompress final
+    {
+        jpeg_error_mgr err{};
+        jpeg_compress_struct enc{};
+
+    public:
+        JPEGCompress() : err{}, enc{}
+        {
+            jpeg_std_error(&err);
+            err.error_exit = &OnJPEGError;
+            enc.err = &err;
+            jpeg_create_compress(&enc);
+        }
+        ~JPEGCompress() noexcept
+        {
+            jpeg_destroy_compress(&enc);
+        }
+
+        void UseOutput(FILE* fout)
+        {
+            jpeg_stdio_dest(&enc, fout);
+        }
+
+        /// @todo More correct DXGI_FORMAT mapping
+        HRESULT WriteImage(const Image& image) noexcept(false)
+        {
+            switch (image.format)
+            {
+            case DXGI_FORMAT_R8_UNORM:
+                enc.input_components = 1;
+                enc.in_color_space = JCS_GRAYSCALE;
+                break;
+            #ifdef LIBJPEG_TURBO_VERSION
+            case DXGI_FORMAT_R8G8B8A8_UNORM:
+                enc.input_components = 4;
+                enc.in_color_space = JCS_EXT_RGBA;
+                break;
+            case DXGI_FORMAT_B8G8R8A8_UNORM:
+                enc.input_components = 4;
+                enc.in_color_space = JCS_EXT_BGRA;
+                break;
+            #else
+            case DXGI_FORMAT_R8G8B8A8_UNORM:
+                enc.input_components = 3;
+                enc.in_color_space = JCS_RGB;
+            #endif
+
+            default:
+                return E_INVALIDARG;
+            }
+            enc.image_width = static_cast<JDIMENSION>(image.width);
+            enc.image_height = static_cast<JDIMENSION>(image.height);
+            jpeg_set_defaults(&enc);
+            jpeg_set_quality(&enc, 100, true);
+            // we will write a row each time ...
+            jpeg_start_compress(&enc, true);
+            const size_t stride = enc.image_width * static_cast<size_t>(enc.input_components);
+            while (enc.next_scanline < enc.image_height)
+            {
+                JSAMPROW rows[1]{ image.pixels + stride * enc.next_scanline };
+                jpeg_write_scanlines(&enc, rows, 1);
+            }
+            jpeg_finish_compress(&enc);
+            return S_OK;
+        }
+    };
+}
+
+_Use_decl_annotations_
+HRESULT DirectX::GetMetadataFromJPEGFile(
+    const wchar_t* file,
+    TexMetadata& metadata)
+{
+    if (!file)
+        return E_INVALIDARG;
+
+    try
+    {
+        auto fin = OpenFILE(file);
+        JPEGDecompress decoder{};
+        decoder.UseInput(fin.get());
+        return decoder.GetHeader(metadata);
+    }
+    catch (const std::system_error& ec)
+    {
+        return ec.code().value();
```

Comment:
I don't believe this is a valid HRESULT. This going to be something from [this list](https://learn.microsoft.com/en-us/cpp/c-runtime-library/errno-constants). We should look at something that translate it back to a HRESULT.

Reply:
I think you want to do is use ``_doserrno`` for Win32:

```
#ifdef _WIN32
    ScopedFILE OpenFILE(const path& p) noexcept(false)
    {
        const std::wstring fpath = p.generic_wstring();
        FILE* fp = nullptr;
        if (auto ec = _wfopen_s(&fp, fpath.c_str(), L"rb"); ec)
            throw std::system_error{ _doserrno, std::system_category(), "_wfopen_s" };
        return { fp, &fclose };
    }
```

Then in the entry-points:

```
    catch (const std::system_error& ec)
    {
        image.Release();
#ifdef _WIN32
        return HRESULT_FROM_WIN32(ec.code().value());
#else
        return E_FAIL;
#endif
    }
```

Add to the include for the .cpp files (official C++ Standard place to find ``errno`` and ``_doserrno``:

```
#include <cstdlib>
```

> For Linux (WSL) there really are only a handful of HRESULTs defined in the adapter headers: ``E_NOTIMPL``, ``E_OUTOFMEMORY ``, ``E_INVALIDARG``, ``E_NOINTERFACE``, ``E_POINTER``, ``E_HANDLE``, ``E_ABORT``, ``E_FAIL``, ``E_ACCESSDENIED``, and ``E_UNEXPECTED``